libraries/standard/mqtt/src/mqtt_lightweight.c

```diff
@@ -190,7 +215,129 @@ static int32_t recvExact( MQTTTransportRecvFunc_t recvFunc,
     return totalBytesRecvd;
 }
 
-/*-----------------------------------------------------------*/
+static bool _publishPacketSize( const MQTTPublishInfo_t * pPublishInfo,
+                                size_t * pRemainingLength,
+                                size_t * pPacketSize )
+{
+    bool status = true;
+    size_t packetSize = 0, payloadLimit = 0;
+
+    /* The variable header of a PUBLISH packet always contains the topic name.
+     * The first 2 bytes of UTF-8 string contains length of the string.
+     */
+    packetSize += pPublishInfo->topicNameLength + sizeof( uint16_t );
+
+    /* The variable header of a QoS 1 or 2 PUBLISH packet contains a 2-byte
+     * packet identifier. */
+    if( pPublishInfo->qos > MQTTQoS0 )
+    {
+        packetSize += sizeof( uint16_t );
+    }
+
+    /* Calculate the maximum allowed size of the payload for the given parameters.
+     * This calculation excludes the "Remaining length" encoding, whose size is not
+     * yet known. */
+    payloadLimit = MQTT_MAX_REMAINING_LENGTH - packetSize - 1U;
+
+    /* Ensure that the given payload fits within the calculated limit. */
+    if( pPublishInfo->payloadLength > payloadLimit )
+    {
+        status = false;
+    }
+    else
+    {
+        /* Add the length of the PUBLISH payload. At this point, the "Remaining length"
+         * has been calculated. */
+        packetSize += pPublishInfo->payloadLength;
+
+        /* Now that the "Remaining length" is known, recalculate the payload limit
+         * based on the size of its encoding. */
+        payloadLimit -= remainingLengthEncodedSize( packetSize );
+
+        /* Check that the given payload fits within the size allowed by MQTT spec. */
+        if( pPublishInfo->payloadLength > payloadLimit )
+        {
+            status = false;
+        }
+        else
+        {
+            /* Set the "Remaining length" output parameter and calculate the full
+             * size of the PUBLISH packet. */
+            *pRemainingLength = packetSize;
+
+            packetSize += 1U + remainingLengthEncodedSize( packetSize );
+            *pPacketSize = packetSize;
+        }
+    }
+
+    return status;
+}
+
+static void _serializePublishCommon( const MQTTPublishInfo_t * pPublishInfo,
+                                     size_t remainingLength,
+                                     uint16_t packetIdentifier,
+                                     const MQTTFixedBuffer_t * const pFixedBuffer,
+                                     bool serializePayload )
+{
+    uint8_t publishFlags = 0;
+    uint8_t * pIndex = pFixedBuffer->pBuffer;
+    size_t minBufferSize = 0;
+
+    /* The first byte of a PUBLISH packet contains the packet type and flags. */
+    publishFlags = MQTT_PACKET_TYPE_PUBLISH;
```

Comment:
Could save a line by initializing it to this

Reply:
Agree.